### PR TITLE
[Inductor] Represent tiling as a dict

### PR DESCRIPTION
**Pull request by blaine-rister**
# Summary

Preparatory refactor for https://github.com/pytorch/pytorch/pull/137243. This makes it easier to generalize to multi-dimensional reductions.

This diff refactors `self.numels` from a tuple like `(8,16)` to a dict like `{"x": 8, "r": 16}`.

Note: this is based off of https://github.com/pytorch/pytorch/pull/141738, which enables `tree.is_reduction`. That PR should land first.

# Test plan
The existing CI provides good coverage.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1439,7 +1440,7 @@ def init_cooperative_reduction(self):
             if tree.grid_dim is not None:
                 tree.grid_dim += 1
 
-        sem_count, _ = self.numels
+        sem_count, _ = self.numels.values()
```
Comment (jansel) [inline torch/_inductor/codegen/triton.py:1443]: This line should be updated to do a dict access to the `x` dim.

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -2470,7 +2473,7 @@ def codegen_cooperative_reduction_peer_combine(self, result_var, dtype):
         column.  After the barrier, every thread block loads the completed value so that it can compute the final
         value independently.
         """
-        xnumel, rnumel = self.numels
+        xnumel, rnumel = self.numels.values()
```
Comment (jansel) [inline torch/_inductor/codegen/triton.py:2476]: This should be updated to do dict access

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/simd.py [hunk 3]
```diff
@@ -1552,7 +1565,23 @@ def candidate_tilings(node):
         return tilings
 
     @classmethod
-    def select_tiling(cls, node_schedule, numel, reduction_numel=sympy.S.One):
+    def create_tiling(
+        cls, pw_tiling: Sequence[sympy.Expr], reduction_tiling: Sequence[sympy.Expr]
+    ) -> Dict[str, Tuple[sympy.Expr]]:
```
Comment (blaine-rister) [inline torch/_inductor/codegen/simd.py:1570]:
```suggestion
    ) -> Dict[str, sympy.Expr]:
```

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (blaine-rister) [reply torch/_inductor/codegen/triton.py]: Updated [d3146d1](https://github.com/pytorch/pytorch/pull/141751/commits/d3146d175c2a6ebbeff582d1561c2539ea1cf91e)

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (blaine-rister) [reply torch/_inductor/codegen/triton.py]: Updated [d3146d1](https://github.com/pytorch/pytorch/pull/141751/commits/d3146d175c2a6ebbeff582d1561c2539ea1cf91e)

**Comment by blaine-rister** [comment]
@atalman I landed the diff internally, but the changes didn't seem to show up in github. There was some mysterious failure preventing GH from landing which seemed like an infra issue. I went ahead and did a "merge -f" to make sure GH and fbcode were synced. Was that the right course of action?

**Comment by Skylion007** [comment]
This also is screaming to me that this should be a TypedDict given that we have a fixed subset of static keys.

**Pull request closed without merging**